[Major Changes][Sprockets 3 for LESS] Using versionized classes

bogdanRada opened the pull request:
Hello.
 I recently added support for Sprockets 3 for sprockets-sass gem ( here: https://github.com/petebrowne/sprockets-sass/pull/38 ) and here https://github.com/petebrowne/sprockets-sass/pull/39  

After some discussion in the first pull request, we decided to use versionized classes, so that would be easier to maintain compatibility with other sprockets versions, so this uses same structure,

Basically the V2 classes will be used for sprockets 2.x
V3 ...for Sprockets 3.x
and V4...for Sprockets 4.x

The Registration class will automatically require the classes based on the version of sprockets loaded.

This pull request will also **DROP Ruby 1.9** and also **DROP TILT DEPENDENCY***

This pull request fixes some issues:
- The importer can now import from globbed files and from deeply nested files
- And now LESS files can be cached, using a custom cache store. 

All tests are now passing for both 2.x and 3.x , except 6 tests related to the helpers we have for assets  ( Which are broken because i honestly haven';t found a way to pass arguments to this methods, besides the first argument )

``` Less
 body { background: asset-path("image.jpg", @prefix: '/themes'); }
```

This seems to thow "Unrecognized input" when trying to parse it with the `Less::Parser` .
 I am going to still try and fix this, but i think this is not available for our `less.js` version (1.6.1) which is loaded by `less.rb`` (version 2.6.0 -- the last version)

There is though LESS.js version 1.7.1, which has support for this, but the `less.js` repository hasn't been updated though :( 

Also there are some changes done for Sprockets 4.x but that is not yet complete. It does not fully support it yet. 

But wanted to get some feedback on this, and maybe we can release the changes for sprockets 3.x and fix those other issues in other pull requests. 

Let me know what you think. Have a nice day :) 

P.S. If you have any idea how to pass those arguments to those helpers i would really appreciate.

bogdanRada committed:
sprockets 3 support

lloeki commented:
Thank you for this, I'll try to look into it soon!

bogdanRada commented:
Thank You too. I have an idea how to fix the issue with the helpers ...but i am going to be out of the country for the next 3 weeks.  Will continue working on that after i return. Thank You for the response and i can't wait to hear more from You :)

bogdanRada commented:
I added few more commits but rebased again and squashed all commits. I fixed the issue with the helpers, now all tests are passing for sprockets 2.x and 3.x . There is still some work left to be done for sprockets 4.x but that can be done in another pull request.

Please review again this pull request and let me know what you think :) Thank you very much

bogdanRada commented:
The tests are passing locally, on Travis they fail because travis somehow uses an older bundler version. Left a comment on this issue https://github.com/travis-ci/travis-ci/issues/5831 . Hopefully someone will find a solution for travis.

bogdanRada commented:
@lloeki , can you please review this pull request again? I fixed some issues with the importer and also the function parsing. I removed also ruby 2.0 and 2.1.5 from Travis YML file since 2.0 is not maintained anymore and 2.1.5 causes issues with bundler (because Travis has a older version of travis on server) 

I also tested this implementation with sprockets 3.x with a real application and works fine . 

Let me know what you think. Thank you very much.

lloeki commented:
I was seeing your commits pouring in, so waited for you to ping me back that you were done :-) I'll look into it soon.

bogdanRada commented:
Thanks :)  , i fixed a minor issue . Replaced "compact.flatten" with "flatten.compact" in lib/sprockets/less/functions.rb. Checked other places. Should be good to go now :) 

Support for Sprockets 4.x is not ready yet, but will continue on that on another pull request.

bogdanRada commented:
Hello. I am just curios :) @loeki did you had any chance to look at this pull request? :)

bogdanRada commented:
The issue with the helpers was fixed. Had to hack it a bit though...but it works. 

Let me know if something needs to Be changed or refactored or something...Thanks a lot :)

lloeki commented:
Wow it'd been so long since I've promised looking at this, sorry! Thanks a lot for your continued effort on this. I have a couple things to do but I'll be looking at it afterwards.

bogdanRada commented:
It's been 5 months since last reply. Any chance of this getting reviewed? Please .

lloeki commented:
Sorry, I'll have a look at it *today*.

lloeki commented on .travis.yml:
```diff
@@ -1,4 +1,25 @@
-  - 2.1.0
+- 2.2.2
+- 2.2.3
+- 2.3.1
```
Can we have 2.4 here too?

bogdanRada replied:
added ruby 2.4.1 to travis

lloeki commented on .travis.yml:
```diff
@@ -1,4 +1,25 @@
-  - 2.1.0
+- 2.2.2
+- 2.2.3
+- 2.3.1
```
Also, can those be specified as major.minor so that we get the latest patch automatically? I can't recall if Travis allows this.

bogdanRada replied:
i don't think Travis has that ability .

lloeki commented on sprockets-less.gemspec:
```diff
@@ -14,10 +14,31 @@ Gem::Specification.new do |s|
+
+  # This is needed so that on rubygems.org we can see the actual date
+  # of the published version due to changes in Rubygems repo over the last year(2015 - 2016)
+  s.date = Date.today
```
I have an issue here with ruby 2.4 (ruby 2.3 plows through fine):

```
[!] There was an error parsing `Gemfile`: 
[!] There was an error while loading `sprockets-less.gemspec`: uninitialized constant Date
```

bogdanRada replied:
That is from bundler actually and i fixed it now.

lloeki commented on lib/sprockets/less.rb:
```diff
@@ -20,5 +26,23 @@ class << self
+    # For Sprockets 2.x , although the file and the module name exist,
+    # they can't be used because it will give errors about undefined methods, because this is included only on Sprockets::Base
+    # and in order to use them we would have to subclass it and define methods to expire cache and other methods for registration ,
+    # which are not needed since Sprockets already  knows about that using the environment instead internally
```
Extra space before `know`. Also, can you hard-wrap this comment to 79?

bogdanRada committed:
add ruby 2.4.1 to travis

lloeki commented on lib/sprockets/less/functions.rb:
```diff
@@ -126,28 +60,92 @@ def add_function(name, &block)
+::Less::Parser.class_eval do
+  attr_reader :tree
+
+
```
Remove that extra line.

bogdanRada committed:
fix bundler removing date

lloeki commented on lib/sprockets/less/functions.rb:
```diff
@@ -126,28 +60,92 @@ def add_function(name, &block)
+      hash[key_value[0].strip.to_sym] = key_value[1].strip
     end
 
+    return hash
```
Remove `return` keyword.

lloeki commented on lib/sprockets/less/functions.rb:
```diff
@@ -126,28 +60,92 @@ def add_function(name, &block)
+       arr << { match: m, index: offset } unless m.nil?
+       str = str[(offset + 1)..-1]
+   end
+   arr.uniq {|hash| hash[:match] }
```
Add blank line before implicit return value

lloeki replied:
Also, missing space after `{`

lloeki commented on lib/sprockets/less/functions.rb:
```diff
@@ -126,28 +60,92 @@ def add_function(name, &block)
+        less.gsub!(function_regex, css)
+      end
+    end
+    less
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/functions.rb:
```diff
@@ -126,28 +60,92 @@ def add_function(name, &block)
+        end
+        params << options  if options.keys.size > 0
+        begin
+        css = @tree.send(@tree.css_to_sym(function_name), *params)
```
Fix indent

lloeki commented on lib/sprockets/less/functions.rb:
```diff
@@ -126,28 +60,92 @@ def add_function(name, &block)
+        captures = hash[:match].captures.map {|a| a.split(',').map {|b| b.gsub(/["']+/, '') } }.flatten.compact.uniq
+        params = []
+        captures.each do |param|
+            if param.include?("@{")
```
Indent by 2, not 4

lloeki commented on lib/sprockets/less/functions.rb:
```diff
@@ -85,11 +9,19 @@ class Tree
+    def extend_js(mod)
+      extend mod
+      mod.public_instance_methods.each do |method_name|
+        add_function(sym_to_css(method_name)) { |tree, cxt|
```
Use `do` / `end` for multiline blocks

lloeki commented on lib/sprockets/less/functions.rb:
```diff
@@ -126,28 +60,92 @@ def add_function(name, &block)
+      function_regex =  /#{function_name}\(([^\)]*)\)/
+      function_data = matches(less, function_regex)
+      function_data.each do |hash|
+        captures = hash[:match].captures.map {|a| a.split(',').map {|b| b.gsub(/["']+/, '') } }.flatten.compact.uniq
```
Space after each blockwise `{`

lloeki commented on lib/sprockets/less/functions.rb:
```diff
@@ -104,15 +36,17 @@ def sym_to_css(sym)
     def unquote(str)
       s = str.to_s.strip
-      s =~ /^['"](.*?)['"]$/ ? $1 : s
+      s.gsub!(/["']+/, '')
```
This removes quotes inside, not just around. Is that what we want?

bogdanRada replied:
yes . That is intended

bogdanRada committed:
fix identations and rake file

lloeki commented on lib/sprockets/less/registration.rb:
```diff
@@ -0,0 +1,113 @@
+# frozen_string_literal: true
+module Sprockets
+  module Less
+    # class useful for registering engines, tranformers, preprocessors and conpressors for sprockets
```
Typo: `compressors`

lloeki commented on lib/sprockets/less/registration.rb:
```diff
@@ -0,0 +1,113 @@
+        require 'sprockets/less/functions'
+      end
+
+    private
```
Align private with `def`/`end`

bogdanRada committed:
fix identations and rake file

lloeki commented on lib/sprockets/less/utils.rb:
```diff
@@ -0,0 +1,137 @@
+        def digest(options)
+          options.delete_if { |_key, value| value.is_a?(Pathname) } if options.is_a?(Hash)
+          options = options.to_s unless options.is_a?(Hash)
+          if defined?(Sprockets::DigestUtils)
```
Add blank line before `if` to materialise implicit return value

lloeki commented on lib/sprockets/less/utils.rb:
```diff
@@ -0,0 +1,137 @@
+              raise Encoding::InvalidByteSequenceError, "#{filename} is not valid #{data.encoding}"
+            end
+          end
+          data
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/utils.rb:
```diff
@@ -0,0 +1,137 @@
+            # Set it to the default external (without verifying)
+            data.force_encoding(Encoding.default_external) if Encoding.default_external
+          end
+          data
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/utils.rb:
```diff
@@ -0,0 +1,137 @@
+        end
+
+        def quote(contents, opts = {})
+          return contents if opts[:unquote]
```
Add blank line after guard clause

lloeki commented on lib/sprockets/less/utils.rb:
```diff
@@ -0,0 +1,137 @@
+          end
+
+          contents = contents.gsub(/\n(?![a-fA-F0-9\s])/, "\\a").gsub("\n", "\\a ")
+          "#{quote}#{contents}#{quote}"
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/utils.rb:
```diff
@@ -0,0 +1,137 @@
+          contents = contents.gsub(/\n(?![a-fA-F0-9\s])/, "\\a").gsub("\n", "\\a ")
+          "#{quote}#{contents}#{quote}"
+        end
+
```
Remove blank line between `end`

lloeki commented on lib/sprockets/less/v2/functions.rb:
```diff
@@ -0,0 +1,147 @@
+          end
+          rv
+        end
+
```
Remove blank line between `end`

lloeki commented on lib/sprockets/less/v2/functions.rb:
```diff
@@ -0,0 +1,147 @@
+            new_key, new_value = yield(k, v)
+            rv[new_key] = new_value
+          end
+          rv
```
Add blank line before implicit return value

bogdanRada committed:
fix travis to install correctly the bundler into vendor folder

lloeki commented on lib/sprockets/less/v2/functions.rb:
```diff
@@ -0,0 +1,147 @@
+
+        def map_hash(hash)
+          # Copy and modify is more performant than mapping to an array and using
+          # to_hash on the result.
```
👍

lloeki commented on lib/sprockets/less/v2/functions.rb:
```diff
@@ -0,0 +1,147 @@
+
+        def fetch_css_url(value ,options = {})
+          new_value = !options[:unquote].nil? ? value : quote(value, options)
+          "url(#{new_value})"
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v2/functions.rb:
```diff
@@ -0,0 +1,147 @@
+      module Functions
+        def asset_data_url(path)
+          value = sprockets_context.asset_data_uri(path)
+          fetch_css_url(value)
```
Add blank line before implicit return value or remove use of `value` variable

lloeki commented on lib/sprockets/less/v2/functions.rb:
```diff
@@ -0,0 +1,147 @@
+        def font_url(source, options = {})
+          # Work with the Compass #font_url API
+          value = font_path(source, options)
+          verify_url_value(value, options)
```
Add blank line before implicit return value or remove use of value variable

lloeki commented on lib/sprockets/less/v2/functions.rb:
```diff
@@ -0,0 +1,147 @@
+
+        def asset_url(path, options = {})
+          value = asset_path(path, options)
+          verify_url_value(value, options)
```
Add blank line before implicit return value or remove use of value variable

lloeki commented on lib/sprockets/less/v2/functions.rb:
```diff
@@ -0,0 +1,147 @@
+
+        def image_path(img, options = {})
+          if options[:from_url] || (options.is_a?(Hash) && options.keys.size > 0)
+           value = quote(sprockets_context.image_path(img, map_options(options)))
```
Indent by 2

lloeki commented on lib/sprockets/less/v2/functions.rb:
```diff
@@ -0,0 +1,147 @@
+
+        def public_path(asset)
+          if sprockets_context.respond_to?(:asset_paths)
+            sprockets_context.asset_paths.compute_public_path asset, defined?(Rails) ? ::Rails.application.config.assets.prefix : '/assets'
```
Parens around args to improve readability

bogdanRada committed:
make travis same as sprockets-sass for consistency -- removed cache for bundler

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+
+        def fetch_all_dependencies(data, filename, less_options, css_options)
+          process_dependencies(data, filename, less_options, css_options)
+          [data, dependencies]
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+            data.gsub!(/^\s*@import\s*['"]#{Regexp.escape(path)}['"]\s*;/, css)
+            final_css << css
+          end
+          final_css
```
Use `each_with_object`, removing need for `final_css` and `import_paths` variables

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+          else
+            css = engine_from_path(path, base_path, less_options, css_options)
+          end
+          css.to_s
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+          rescue => e
+            css = e.message =~ /is undefined/ ? data_to_parse : nil
+          end
+          css.nil? || css.empty? ? data_to_parse : css
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+          css.nil? || css.empty? ? data_to_parse : css
+        end
+
+
```
Remove extra blank line

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+            ))
+            engine = parser.parse(data)
+          end
+          engine.to_css(css_options)
```
Add blank line before implicit return value

bogdanRada committed:
forgot the no-deployment flag for travis

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+            importer: self,
+            custom: { sprockets_context: context }
+            ))
+            engine = parser.parse(data)
```
Do something like 

```
less_options = less_options.merge(...)
engine = ::Less::Parser.new(less_options).parse(data)
```

Thus removing the need for double `))`

bogdanRada committed:
ident travis conf file

lloeki commented on spec/support/less_template.rb:
```diff
@@ -0,0 +1,7 @@
+  module Less
+    class LessTemplate < Sprockets::Less::Utils.get_class_by_version("LessTemplate")
+
+    end
```
Use `class ...; end` style for empty classes

lloeki commented on spec/support/test_registration.rb:
```diff
@@ -0,0 +1,11 @@
+      def register_engines(hash)
+        _register_engines(hash)
+      end
+
```
Remove blank line between `end`

lloeki commented on spec/support/test_registration.rb:
```diff
@@ -0,0 +1,11 @@
+module Sprockets
+  module Less
+    class TestRegistration < Sprockets::Less::Registration
+
```
Remove blank line

lloeki commented on sprockets-less.gemspec:
```diff
@@ -1,7 +1,7 @@
 $:.push File.expand_path("../lib", __FILE__)
 require "sprockets/less/version"
-
+require 'date'
```
Keep blank line after `require` block.

lloeki commented on spec/support/fake_engine.rb:
```diff
@@ -0,0 +1,50 @@
+          result
+        end
+      end
+
```
Remove blank line between `end`

lloeki commented on spec/support/fake_engine.rb:
```diff
@@ -0,0 +1,50 @@
+      def run
+        @has_been_used = true
+        result = ""
+        if context.respond_to?(:metadata)
```
Add blank line before implicit return value

lloeki commented on spec/support/fake_engine.rb:
```diff
@@ -0,0 +1,50 @@
+      end
+
+      def call(input)
+        @context  = input[:environment].context_class.new(input)
```
Remove extra space before `=`

lloeki commented on spec/support/fail_postprocessor.rb:
```diff
@@ -0,0 +1,23 @@
+    context  = input[:environment].context_class.new(input)
+
+    result = run(filename, source, context)
+    context.metadata.merge(data: result)
```
Add blank line before implicit return value

lloeki commented on spec/support/dummy_importer.rb:
```diff
@@ -0,0 +1,25 @@
+
+      def find(path, options)
+        @has_been_used = true
+        super(path, options)
```
Add blank line before implicit return value

lloeki commented on spec/support/dummy_importer.rb:
```diff
@@ -0,0 +1,25 @@
+      
+      def find_relative(uri, base, options, *args)
+        @has_been_used = true
+        super(uri, base, options, *args)
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v4/importer.rb:
```diff
@@ -0,0 +1,108 @@
+        def filtered_processor_classes
+          classes = super
+          classes << Sprockets::YUICompressor if defined?(Sprockets::YUICompressor)
+          classes
```
I fancy to use `super.tap { |classes| classes << Sprockets::YUICompressor if defined?(Sprockets::YUICompressor) }` in those situations. Not a required change though.

lloeki commented on lib/sprockets/less/v4/importer.rb:
```diff
@@ -0,0 +1,108 @@
+          if processor.respond_to?(:processors)
+            processor.processors = filter_all_processors(processor.processors)
+          end
+          super(processor, context, input, processors)
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v4/importer.rb:
```diff
@@ -0,0 +1,108 @@
+          css_transformers = available_transformers.key?('text/css') ? available_transformers['text/css'] : []
+          css_transformers = css_transformers.is_a?(Array) ? css_transformers : [css_transformers]
+          additional_transformers = additional_transformers.concat(css_transformers)
+          additional_transformers
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v4/importer.rb:
```diff
@@ -0,0 +1,108 @@
+          pathname = context.resolve(path)
+          content_type = pathname.nil? ? nil : pathname.to_s.scan(/\?type\=(.*)/).flatten.compact.first unless pathname.nil?
+          attributes = {}
+          [content_type, attributes]
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v4/importer.rb:
```diff
@@ -0,0 +1,108 @@
+
+        def stat_of_pathname(context, pathname, _path)
+          filename = check_path_before_process(context, pathname)
+          context.environment.stat(filename)
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v4/importer.rb:
```diff
@@ -0,0 +1,108 @@
+        def possible_files(context, path, base_path)
+          filename = check_path_before_process(context, base_path)
+          base_path = (filename.is_a?(Pathname) ? filename : Pathname.new(filename))
+          super(context, path, base_path)
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v4/importer.rb:
```diff
@@ -0,0 +1,108 @@
+            asset_requirable = asset_requirable?(context, pathname)
+            imports << { file_url: pathname, path: path } if path != context.filename && asset_requirable
+          end
+          glob_files
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v4/importer.rb:
```diff
@@ -0,0 +1,108 @@
+            imports << %(@import "#{relative_path}";\n)
+          end
+          return nil if engine_imports.empty?
+          ::Sass::Engine.new engine_imports, options.merge(
```
Add blank line before implicit return value

bogdanRada replied:
THe v4 importer is not ready yet, i copied it from sprockets-sass aparently, I need to replace Sass::Engine with Less::Parser,   I still need to refactor this class completely, I almost fixed the v4 importer for SASS...i have 2 tests that don't want to run there, but i will try to make this work too.

lloeki commented on lib/sprockets/less/v4/importer.rb:
```diff
@@ -0,0 +1,108 @@
+            relative_path = path[:path].relative_path_from Pathname.new(base_path).dirname
+            imports << %(@import "#{relative_path}";\n)
+          end
+          return nil if engine_imports.empty?
```
Add blank line before guard clause

lloeki commented on lib/sprockets/less/v4/cache_store.rb:
```diff
@@ -0,0 +1,11 @@
+    module V4
+      # Internal: Cache wrapper for Sprockets cache adapter. (Sprockets >= 3)
+      class CacheStore < Sprockets::Less::V3::CacheStore
+      end
```
Use `class ...; end` style for empty classes

lloeki commented on lib/sprockets/less/v3/less_template.rb:
```diff
@@ -0,0 +1,33 @@
+      # Preprocessor for SASS files
+      class LessTemplate < Sprockets::Less::V2::LessTemplate
+        def build_cache_store(context)
+          return nil if context.environment.cache.nil?
```
Add blank line after guard clause

lloeki commented on lib/sprockets/less/v3/less_template.rb:
```diff
@@ -0,0 +1,33 @@
+          return nil if context.environment.cache.nil?
+          cache = @input[:cache]
+          version = @cache_version
+          custom_cache_store(cache, version)
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v3/less_template.rb:
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+require_relative '../v2/less_template'
```
Add blank line after `require` block

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+          filter_all_processors(processors)
+          evaluate_path_from_context(context, path, processors)
+        end
+
```
Remove blank line between `end`

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+             ext_engines = ext_engines.is_a?(Array) ? ext_engines : [ext_engines]
+             engines.concat(ext_engines)
+           end
+           engines
```
Add blank line before implicit return value

lloeki replied:
Wait, rather, use `each_with_object`

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+              data += process(processors, context, file_asset.filename)
+            end
+          end
+          data
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+            else
+              raise Error, "invalid processor return type: #{result.class}"
+          end
+          data
```
~~Add blank line before implicit return value~~

lloeki replied:
Hmm. We should just skip the `data` var assignment altogether and use `case ... end` as a return value.

bogdanRada committed:
travis fix

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+          if result.key?(:required)
+            result[:required].each do |file|
+              file_asset = context.environment.load(file)
+              data += process(processors, context, file_asset.filename)
```
Can `result.key?(:data)` be false when `result.key?(:required)` is true? See above comment.

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+        end
+
+        def handle_complex_process_result(context, result, processors)
+          data = result[:data] if result.key?(:data)
```
Should this be `result.key?(:required)`? Maybe this can be factored into the next `if`.

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+          metadata[:data] = input[:data]
+          result = processor.call(input)
+          processors.delete(processor)
+          handle_process_result(context, result, processors, metadata)
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+
+        def build_input_for_process(context, path, data)
+          content_type, _attributes = content_type_of_path(context, path)
+          {
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+        def content_type_of_path(context, path)
+          attributes = context.environment.send(:parse_path_extnames, path.to_s)
+          content_type = attributes[1]
+          [content_type, attributes]
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+          classes = super
+          classes << Sprockets::Preprocessors::DefaultSourceMap if defined?(Sprockets::Preprocessors::DefaultSourceMap)
+          classes << Sprockets::SourceMapProcessor if defined?(Sprockets::SourceMapProcessor)
+          classes
```
`super.tap do |classes| ... end` is a possibility.

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+            paths.unshift(relative_path, partialize_path(relative_path))
+          end
+          paths = paths.unshift(relatives)
+          paths
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+        def possible_files(context, path, base_path)
+          paths, root_path = super(context, path, base_path)
+          paths = additional_paths_for_sprockets(context, paths, path, base_path)
+          [paths.compact, root_path]
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+          return false if pathname.nil?
+          stat = stat_of_pathname(context, pathname, path)
+          return false unless stat && stat.file?
+          check_context_content_types(context, path)
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+        def check_context_content_types(context, path)
+          available_mimes = available_content_types(path)
+          path_content_type, _attributes = content_type_of_path(context, path)
+          context.content_type.nil? || available_mimes.include?(path_content_type) || available_mimes.include?(context.content_type)
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+            found_item = resolve_path_with_load_paths(context, path, root_path, file)
+            return found_item if !found_item.nil? && asset_requirable?(context, found_item)
+          end
+          nil
```
Use `Enumerable#find` instead of `each`

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+      # class used for importing files from LES files
+      class Importer < Sprockets::Less::V2::Importer
+
+
```
Remove extra blank line

lloeki commented on lib/sprockets/less/v2/less_template.rb:
```diff
@@ -0,0 +1,252 @@
+          cache_store: build_cache_store(context),
+          importer: importer,
+          custom: { sprockets_context: context },
+          sprockets: sprockets_options
```
Indent by 2

lloeki commented on lib/sprockets/less/v2/less_template.rb:
```diff
@@ -0,0 +1,252 @@
+          custom: { sprockets_context: context },
+          sprockets: sprockets_options
+          )
+          less
```
Remove `less` variable

lloeki commented on lib/sprockets/less/v2/less_template.rb:
```diff
@@ -0,0 +1,252 @@
+
+        def store_into_cache_store(data)
+          return if less_options[:cache] == false || less_options[:cache_store].nil?
+          less_options[:cache_store]._store(@cache_key, @cache_version,  cache_sha_filename, data)
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v2/less_template.rb:
```diff
@@ -0,0 +1,252 @@
+
+        def retrieve_from_cache_store
+          return if less_options[:cache] == false || less_options[:cache_store].nil?
+          less_options[:cache_store]._retrieve(@cache_key, @cache_version,  cache_sha_filename)
```
Add blank line before implicit return value

bogdanRada committed:
another attempt to make travis work

lloeki commented on lib/sprockets/less/v2/less_template.rb:
```diff
@@ -0,0 +1,252 @@
+        end
+
+        def default_less_options
+          default_less_config[:load_paths] =  default_less_config[:load_paths].is_a?(Array) ? default_less_config[:load_paths] : []
```
Remove extra space

lloeki commented on lib/sprockets/less/v2/less_template.rb:
```diff
@@ -0,0 +1,252 @@
+        def default_less_options
+          default_less_config[:load_paths] =  default_less_config[:load_paths].is_a?(Array) ? default_less_config[:load_paths] : []
+          default_less_config[:load_paths] = default_less_config[:load_paths].concat(default_less_config[:paths]) if default_less_config[:paths].is_a?(Array)
+          less =  default_less_config
```
Remove extra space

lloeki commented on lib/sprockets/less/v2/less_template.rb:
```diff
@@ -0,0 +1,252 @@
+          default_less_config[:load_paths] = default_less_config[:load_paths].concat(default_less_config[:paths]) if default_less_config[:paths].is_a?(Array)
+          less =  default_less_config
+          less = merge_less_options(less.dup, @less_config) if defined?(@less_config) && @less_config.is_a?(Hash)
+          less
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v2/less_template.rb:
```diff
@@ -0,0 +1,252 @@
+          options[:load_paths] = options[:load_paths].is_a?(Array) ? options[:load_paths] : []
+
+          if (load_paths = options[:paths]) && (other_paths = other_options[:paths])
+            options[:load_paths] = options[:load_paths]+ other_paths + load_paths
```
Missing space around `+`

lloeki commented on lib/sprockets/less/v2/less_template.rb:
```diff
@@ -0,0 +1,252 @@
+          fetch_importer_class.fetch_all_dependencies(data, filename, less_options, css_options)
+        end
+
+        def merge_less_options(options, other_options)
```
So this is side-effectful on both `options` and `other_options`? I'd have expected only `options` to suffer from side effects (like `Hash#merge!`). Maybe some `dup` is in order, or explain the rationale and/or behaviour in a comment.

lloeki commented on lib/sprockets/less/v2/less_template.rb:
```diff
@@ -0,0 +1,252 @@
+            options[:load_paths] = options[:load_paths]+ other_paths + load_paths
+          end
+          options[:load_paths] = options[:load_paths].concat(context.environment.paths)
+          options
```
Remove extra spaceAdd blank line before implicit return value

lloeki commented on lib/sprockets/less/v2/less_template.rb:
```diff
@@ -0,0 +1,252 @@
+
+        def css_options
+          css_keys = [:compress, :optimization, :silent, :color]
+          Hash[less_options.to_enum.to_a.select{|k, _| css_keys.include? k}]
```
Add blank line before implicit return value

lloeki replied:
spaces all around block's `{` and `}`

lloeki commented on lib/sprockets/less/v2/less_template.rb:
```diff
@@ -0,0 +1,252 @@
+          else
+            css
+          end
+
```
Remove blank line

lloeki commented on lib/sprockets/less/v2/less_template.rb:
```diff
@@ -0,0 +1,252 @@
+            data = Sprockets::Less::Utils.read_file_binary(filename, options)
+            new_data, dependencies = process_dependencies(data)
+
+            css  = less_engine(new_data, less_options, css_options)
```
Remove extra space before `=`

lloeki commented on lib/sprockets/less/v2/less_template.rb:
```diff
@@ -0,0 +1,252 @@
+            end
+          else
+            css
+          end
```
Line 135 exit point is easily overlooked due to too much nesting. Turn `if css.nil?` and `if context.respond_to?(:metadata)` into guard clauses that `return css`.

lloeki commented on lib/sprockets/less/v2/less_template.rb:
```diff
@@ -0,0 +1,252 @@
+          run
+        end
+
+
```
Remove extra blank line

lloeki commented on lib/sprockets/less/v2/less_template.rb:
```diff
@@ -0,0 +1,252 @@
+          @filename = input[:filename]
+          @source   = input[:data]
+          @context  = input[:environment].context_class.new(input)
+          run
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+          engines = get_engines_from_attributes(context, attributes)
+          preprocessors = get_context_preprocessors(context, content_type)
+          additional_transformers = get_context_transformers(context, content_type, path)
+          additional_transformers.reverse + preprocessors + engines.reverse
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+        def content_type_of_path(context, path)
+          attributes = context.environment.attributes_for(path)
+          content_type = attributes.content_type
+          [content_type, attributes]
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+        def filtered_processor_classes
+          classes = [Sprockets::Less::Utils.get_class_by_version('LessTemplate')]
+          classes << Tilt::LessTemplate if defined?(Tilt::LessTemplate)
+          classes
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+        # Returns nil if the path is already to a partial.
+        def partialize_path(path)
+          return unless path.basename.to_s !~ /\A_/
+          Pathname.new path.to_s.sub(/([^\/]+)\Z/, '_\1')
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+            relative_path = base_path.relative_path_from(root_path).join path
+            paths.unshift(relative_path, partialize_path(relative_path))
+          end
+          [paths, root_path]
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+
+          # Find base_path's root
+          paths, root_path = add_root_to_possible_files(context, base_path, path, paths)
+          [paths.compact, root_path]
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+              return found if context.asset_requirable?(found)
+            end
+          end
+          nil
```
Use `Enumerable#find` instead of `each`

bogdanRada commented:
Travis currently is failing , there seem to be some kind of caching problem, i am still trying to figure that out. Seems strange though, for other repositories i use same configuration, but for this repository it doesn't want to work . Need to spend more time investigating this problem

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+          context = less_options[:custom][:sprockets_context]
+          (pathname = resolve(context, path, base_path))  || (return nil)
+          context.depend_on pathname
+          dependencies << pathname
```
Add blank line after to make the structure more clear

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+              css = find_relative(import_path, pathname, less_options, css_options)
+              data_to_parse.gsub!(/^\s*@import\s*['"]#{Regexp.escape(import_path)}['"]\s*;/, css)
+            end
+          end
```
Add blank line after to make the structure more clear

lloeki commented on lib/sprockets/less/v2/importer.rb:
```diff
@@ -0,0 +1,250 @@
+          final_css
+        end
+
+
```
Remove extra blank line

lloeki commented on lib/sprockets/less/v2/functions.rb:
```diff
@@ -0,0 +1,147 @@
+
+        def font_path(source, options = {})
+          if options[:from_url] || (options.is_a?(Hash) && options.keys.size > 0)
+           quote(sprockets_context.font_path(source, map_options(options)))
```
Indent by 2

lloeki commented on lib/sprockets/less/v2/cache_store.rb:
```diff
@@ -0,0 +1,28 @@
+          @environment = environment
+        end
+
+
```
Remove extra blank line

lloeki commented on lib/sprockets/less/v2/cache_store.rb:
```diff
@@ -0,0 +1,28 @@
+          return if obj[:version] != version || obj[:sha] != sha
+          obj[:obj]
+        end
+
```
Remove blank line between `end`

lloeki commented on lib/sprockets/less/v2/cache_store.rb:
```diff
@@ -0,0 +1,28 @@
+          obj = environment.send(:cache_get, "less/#{key}")
+          return unless obj.is_a?(Hash)
+          return if obj[:version] != version || obj[:sha] != sha
+          obj[:obj]
```
Add blank line before implicit return value

lloeki commented:
Wow that was one hell of a review, nice piece of work, thanks a lot! I didn't find much questionable things :D Mostly that's about some style issues.

So let me know if you can find anything on the Travis front, fix the nitpicks and we can finally get this merged!

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+          preprocessors = get_context_preprocessors(context, content_type)
+          additional_transformers = get_context_transformers(context, content_type, path)
+          postprocessors = get_context_postprocessors(context, content_type)
+          engines.reverse + preprocessors + additional_transformers.reverse  + postprocessors
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+          content_type, attributes = content_type_of_path(context, path)
+          processors = get_all_processors_for_evaluate(context, content_type, attributes, path)
+          filter_all_processors(processors)
+          evaluate_path_from_context(context, path, processors)
```
Add blank line before implicit return value

lloeki commented on lib/sprockets/less/v3/importer.rb:
```diff
@@ -0,0 +1,224 @@
+        def get_context_transformers(context, content_type, path)
+          available_transformers =  context.environment.transformers[content_type]
+          additional_transformers = available_transformers.key?(syntax_mime_type) ? available_transformers[syntax_mime_type] : []
+          additional_transformers.is_a?(Array) ? additional_transformers : [additional_transformers]
```
Add blank line before implicit return value

lloeki commented on sprockets-less.gemspec:
```diff
@@ -14,10 +14,31 @@ Gem::Specification.new do |s|
-  s.add_development_dependency 'rake'
+  s.add_development_dependency 'appraisal', '~> 2.1', '>= 2.1'
+  s.add_development_dependency 'rake', '>= 10.5', '>= 10.5'
 end
```
Add `rubocop` as a version-softlocked development dependency:

```
  s.add_development_dependency 'rubocop', '~> 0.46.0'
```
